lang/cs/Org.Apache.REEF.Network/Elastic/Failures/Default/DefaultFailureStateMachine.cs

```diff
@@ -0,0 +1,320 @@
+
+                float currentRate = (float)NumOfFailedDataPoints / NumOfDataPoints;
+
+                if (isFinalState.Contains(State.FailureState) && 
+                    currentRate >= transitionWeights[DefaultFailureStates.StopAndReschedule])
+                {
+                    throw new IllegalStateException("Received remove data point when state is complete: failing.");
+                }
+
+                while (State.FailureState < (int)DefaultFailureStates.Fail &&
+                    currentRate > transitionWeights[transitionMapUp[(DefaultFailureStates)State.FailureState]])
+                {
+                    State.FailureState = (int)transitionMapUp[(DefaultFailureStates)State.FailureState];
+                }
+
+                return State;
+            }
+        }
+
+        /// <summary>
+        /// Signal the state machine to move into complete state.
+        /// </summary>
+        public IFailureState Complete()
+        {
+            lock (_statusLock)
+            {
+                if (canMoveToComplete.Contains(State.FailureState))
+                {
+                    State.FailureState = (int)DefaultFailureStates.Complete;
+                }
+                else
+                {
+                    throw new IllegalStateException(
+                        $"Failure machine cannot move from state {State.FailureState} to Complete: failing.");
+                }
+            }
+
+            return State;
+        }
+
+        /// <summary>
+        /// Method used to set or update the current threshold connected with
+        /// a target failure state. The assumption is that higher failure states
+        /// have higher thresholds.
+        /// </summary>
+        /// <param name="level">The failure state we want to change</param>
+        /// <param name="threshold">A [0, 1] value specifying when the failure level is reached</param>
+        public void SetThreshold(IFailureState level, float threshold)
+        {
+            if (!(level is DefaultFailureState))
+            {
+                throw new ArgumentException(level.GetType() + " is not DefaultFailureStateMachine");
+            }
+
+            if (level.FailureState == (int)DefaultFailureStates.Continue)
+            {
+                throw new ArgumentException("Cannot change the threshold for Continue state");
+            }
+
+            lock (_statusLock)
+            {
+                transitionWeights[(DefaultFailureStates)level.FailureState] = threshold;
+
+                CheckConsistency();
+            }
+        }
+
+        /// <summary>
+        /// A utility method for setting multiple threshold at once.
+        /// </summary>
+        /// <param name="weights">Pairs of failure states with related new thresholds</param>
+        public void SetThresholds(params Tuple<IFailureState, float>[] weights)
+        {
+            if (!weights.All(weight => weight.Item1 is DefaultFailureState))
```

Comment:
`.Any()` is better than `.All()` in this case, i.e.
```C#
if (weights.Any(weight => !(weight.Item1 is DefaultFailureState)))
```
better yet, we can do both checks in one pass (if we don't want to do it in the last `foreach`):
```C#
foreach (var weight in weights)
{
    if (!(weight.Item1 is DefaultFailureState))
    {
        throw new ArgumentException("Input is not of type DefaultFailureStateMachine,");
    }

    if (weight.Item1.FailureState == (int)DefaultFailureStates.Continue)
    {
        throw new ArgumentException("Cannot change the threshold for Continue state.");
    }
}
```

Reply:
Really good point. Fixed.